src/index.js

```diff
@@ -301,44 +301,50 @@ class TerserPlugin {
 
           // Write extracted comments to commentsFile
           if (commentsFile && extractedComments.length > 0) {
-            // Add a banner to the original file
-            if (this.options.extractComments.banner !== false) {
-              let banner =
-                this.options.extractComments.banner ||
-                `For license information please see ${path.posix.basename(
-                  commentsFile
-                )}`;
-
-              if (typeof banner === 'function') {
-                banner = banner(commentsFile);
-              }
+            // Filter out duplicate comments, if commentsFile already exits
+            let dedupedComments = extractedComments;
+            if (commentsFile in compilation.assets) {
+              const commentsFileSource = compilation.assets[
+                commentsFile
+              ].source();
 
-              if (banner) {
-                outputSource = new ConcatSource(
-                  `/*! ${banner} */\n`,
-                  outputSource
-                );
-              }
+              dedupedComments = dedupedComments.filter(
+                (comment) => !commentsFileSource.includes(comment)
+              );
             }
 
-            const commentsSource = new RawSource(
-              `${extractedComments.join('\n\n')}\n`
-            );
+            if (dedupedComments.length > 0) {
+              // Add a banner to the original file
+              if (this.options.extractComments.banner !== false) {
+                let banner =
+                  this.options.extractComments.banner ||
+                  `For license information please see ${path.posix.basename(
+                    commentsFile
+                  )}`;
+
+                if (typeof banner === 'function') {
+                  banner = banner(commentsFile);
+                }
+
+                if (banner) {
+                  outputSource = new ConcatSource(
+                    `/*! ${banner} */\n`,
+                    outputSource
+                  );
+                }
+              }
 
-            if (commentsFile in compilation.assets) {
-              // commentsFile already exists, append new comments...
-              if (compilation.assets[commentsFile] instanceof ConcatSource) {
+              const commentsString = `${dedupedComments.join('\n\n')}\n`;
```

Comment:
Please use `RawSource` here instead `string` (better use `webpack-default` always - better compatibility and perf).

Reply:
Hmm, according to https://github.com/webpack/webpack-sources#concatsource `ConcatSource`'s constructor and `.add()` method accept either source or string. So I figured it would be better for performance to not create additional `RawSource ` instance since it will be appended to `ConcatSource` anyway.